common/functions/src/scalars/strings/substring_index.rs

```diff
@@ -0,0 +1,261 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use common_datavalues::prelude::*;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use itertools::izip;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct SubstringIndexFunction {
+    display_name: String,
+}
+
+impl SubstringIndexFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(SubstringIndexFunction {
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+}
+
+impl Function for SubstringIndexFunction {
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn num_arguments(&self) -> usize {
+        3
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        if !args[0].is_numeric() && args[0] != DataType::String && args[0] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected string or null, but got {}",
+                args[0]
+            )));
+        }
+        if !args[1].is_numeric() && args[1] != DataType::String && args[1] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[1]
+            )));
+        }
+        if !args[2].is_integer() && args[2] != DataType::String && args[2] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[2]
+            )));
+        }
+        Ok(DataType::String)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
+        let s_column = columns[0].column().cast_with_type(&DataType::String)?;
+        let d_column = columns[1].column().cast_with_type(&DataType::String)?;
+        let c_column = columns[2].column().cast_with_type(&DataType::Int64)?;
+
+        let r_column: DataColumn = match (s_column, d_column, c_column) {
```

Comment:
That's great.

But I have some thoughts, how about forcing the third argument to be Constant?
I did not find any reason to use dynamic count.

Thus we can reduce lots of branches.

Reply:
From ClickHouse doc:

https://github.com/ClickHouse/ClickHouse/blob/cf0fbbbc1704e8d232039547005424830a0c82d9/docs/en/sql-reference/functions/string-functions.md